programs/uxd/src/instructions/mercurial/collect_profit_of_mercurial_vault_depository.rs

```diff
@@ -0,0 +1,298 @@
+        collateral_mint.as_ref(),
+        &[depository_bump],
+    ]];
+
+    let before_lp_token_vault_balance = ctx.accounts.depository_lp_token_vault.amount;
+    let before_collateral_balance = ctx.accounts.profits_beneficiary_collateral.amount;
+
+    // 1 - calculate the value of collectable interests and fees (in USDC unit)
+    let collectable_profits_value = ctx.accounts.calculate_collectable_profits_value()?;
+
+    // 2 - calculate the amount of LP token to withdraw to collect all interests and fees
+    // This LP amount is subjected to precision loss (we handle this precision loss later)
+    let lp_token_amount_to_match_collectable_profits_value = ctx
+        .accounts
+        .calculate_lp_token_amount_to_match_collectable_profits_value(collectable_profits_value)?;
+
+    let possible_lp_token_precision_loss_collateral_value =
+        mercurial_utils::calculate_possible_lp_token_precision_loss_collateral_value(
+            &ctx.accounts.mercurial_vault,
+            ctx.accounts.mercurial_vault_lp_mint.supply,
+        )?;
+
+    // 3 - withdraw collateral from mercurial vault for LP tokens
+    mercurial_vault::cpi::withdraw(
+        ctx.accounts
+            .into_withdraw_collateral_from_mercurial_vault_context()
+            .with_signer(depository_signer_seed),
+        lp_token_amount_to_match_collectable_profits_value,
+        // Do not check slippage here
+        0,
+    )?;
+
+    // 4 - Reload accounts impacted by the withdraw (We need updated numbers for further calculation)
+    ctx.accounts.depository_lp_token_vault.reload()?;
+    ctx.accounts.profits_beneficiary_collateral.reload()?;
+
+    // 5 - Check that a positive amount of collateral have been redeemed
+    let after_collateral_balance = ctx.accounts.profits_beneficiary_collateral.amount;
+
+    let collateral_balance_change = after_collateral_balance
+        .checked_sub(before_collateral_balance)
+        .ok_or_else(|| error!(UxdError::MathError))?;
+
+    require!(
+        collateral_balance_change > 0,
+        UxdError::MinimumRedeemedCollateralAmountError
+    );
+
+    // 6 - Check the amount of paid LP Token when calling mercurial_vault::cpi::withdraw
+    let after_lp_token_vault_balance = ctx.accounts.depository_lp_token_vault.amount;
+
+    let lp_token_change = before_lp_token_vault_balance
+        .checked_sub(after_lp_token_vault_balance)
+        .ok_or_else(|| error!(UxdError::MathError))?;
+
+    require!(
+        lp_token_change == lp_token_amount_to_match_collectable_profits_value,
+        UxdError::SlippageReached,
+    );
+
+    // 7 - Check the collateral amount received from mercurial_vault::cpi::withdraw
+
+    // There can be precision loss when calculating how many LP token to withdraw and also when withdrawing the collateral
+    // We accept theses losses as the money is still in the vault. We collect a bit less profit.
+    CollectProfitOfMercurialVaultDepository::check_redeemed_collateral_amount_to_match_target(
```

Comment:
nit: this function is the same as the Redeem IX, we could make a utility function for code re-use

Reply:
contents are the same, but I guess the naming here should be check_collected_collateral... sth,
anyway shd be a util func and hv some more general naming